### PR TITLE
[Sparkclient] Adding the sparkclient runtime

### PR DESCRIPTION
**Pull request by shul**

### CODE REVIEW

File: mlrun/runtimes/sparkclientjob.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    def spec(self, spec):
+        self._spec = self._verify_dict(spec, "spec", SparkClientSpec)
+
+    def with_igz_spark(self, spark_service):
+        self.spec.igz_spark = True
+        self.spec.env.append({"name": "MLRUN_SPARK_CLIENT_IGZ_SPARK", "value": "true"})
+        self.apply(mount_v3io_extended())
+        self.apply(
+            mount_v3iod(
+                namespace="default-tenant",
```
Comment (urihoenig) [inline mlrun/runtimes/sparkclientjob.py:79]: Not urgent in this PR IMO (we have the same issue in sparkjob), but if possible, we should get the right namespace here.
Not sure it's easy as there's still no runobj at this stage. @Hedingber ?

File: mlrun/runtimes/sparkclientjob.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        if not self.spec.build.base_image:
+            self.spec.build.base_image = self._default_image
+        return super().deploy(
+            watch=watch,
+            with_mlrun=with_mlrun,
+            skip_deployed=skip_deployed,
+            is_kfp=is_kfp,
+        )
+
+    def _run(self, runobj: RunObject, execution: MLClientCtx):
```
Comment (urihoenig) [inline mlrun/runtimes/sparkclientjob.py:120]: As we don't set the image name here, this can be removed.

File: mlrun/runtimes/sparkclientjob.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    @staticmethod
+    def _get_object_label_selector(object_id: str) -> str:
+        return f"mlrun/uid={object_id}"
+
+    @staticmethod
+    def _get_default_label_selector() -> str:
+        return "mlrun/class=sparkclient"
+
+
+def igz_spark_pre_hook():
```
Comment (urihoenig) [inline mlrun/runtimes/sparkclientjob.py:140]: Not sure, it could be a static method inside the runtime, but I'm not sure it's right

**Review by urihoenig — COMMENTED** [review]
I haven't reviewed the tests.
As for the rest, I've left some comments.
Documentation & example are still missing (can be in another PR).
General thoughts - what is sparkclient without igz_spark=True? Is it just a kubejob?
Should we rename it to be igzsparkclientjob.py, remove the spec and act as if igz_spark is always true? @Hedingber - Your thoughts?

File: mlrun/runtimes/sparkclientjob.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            command=command,
+            args=args,
+            image=image,
+            mode=mode,
+            build=build,
+            entry_points=entry_points,
+            description=description,
+            workdir=workdir,
+            default_handler=default_handler,
+        )
```
Comment (Hedingber) [inline mlrun/runtimes/sparkclientjob.py:58]: There are a lot of things you're not passing to the base class, why ?

File: mlrun/runtimes/sparkclientjob.py (hunk 1 above)
Comment (Hedingber) [reply mlrun/runtimes/sparkclientjob.py:79]: You have it in the config, wasn't aware it's hard coded in sparkjob, please fix it there as well

File: mlrun/runtimes/sparkclientjob.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            )
+        )
+
+    @property
+    def is_deployed(self):
+        if (
+            not self.spec.build.source
+            and not self.spec.build.commands
+            and not self.spec.build.extra
+            and self.spec.igz_spark
```
Comment (Hedingber) [inline mlrun/runtimes/sparkclientjob.py:90]: Are you sure the `igz_spark` is relevant here ? if so add a comment explaining why ?

File: mlrun/runtimes/sparkclientjob.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        ):
+            app_image = re.sub("spark-app", "shell", config.spark_app_image)
+            # this is temporary until we get the image name from external config
+            return app_image + ":" + config.spark_app_image_tag
+        return None
+
+    def deploy(self, watch=True, with_mlrun=True, skip_deployed=False, is_kfp=False):
+        """deploy function, build container with dependencies"""
+        # connect will populate the config from the server config
+        get_run_db()
```
Comment (Hedingber) [inline mlrun/runtimes/sparkclientjob.py:110]: Not needed anymore, remove

File: mlrun/runtimes/sparkclientjob.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            not self.spec.build.source
+            and not self.spec.build.commands
+            and not self.spec.build.extra
+            and self.spec.igz_spark
+        ):
+            return True
+        return super().is_deployed
+
+    @property
+    def _default_image(self):
```
Comment (Hedingber) [inline mlrun/runtimes/sparkclientjob.py:96]: Change to a function named `_resolve_default_base_image()`

**Review by Hedingber — CHANGES_REQUESTED** [review]

File: mlrun/runtimes/sparkclientjob.py (hunk 6 above)
Comment (urihoenig) [reply mlrun/runtimes/sparkclientjob.py:110]: @Hedingber Does it mean that config will be filled on the client side as soon as we use new_function?

Comment (urihoenig) [reply mlrun/runtimes/sparkclientjob.py:110]: Does it mean that it won't be needed in with_igz_spark too? (In order to get the namespace)

Comment (Hedingber) [reply mlrun/runtimes/sparkclientjob.py]: It's done as soon as you `import mlrun` or setting the `config.dbpath` so yeah not needed anywhere

**Comment by Hedingber** [comment]
@urihoenig will continue the work on it in https://github.com/mlrun/mlrun/pull/757
Closing this one

**Pull request closed without merging**